### ISSUE TITLE
Issue with integration with self-signed-certificates since revision 174 on latest/stable

### ISSUE THREAD

**Issue by Thanhphan1147** [canonical/traefik-k8s-operator#322]
### Bug Description

Hi, thanks for your work on traefik!

I noticed that since revision 174 is promoted to `latest/stable`, integrations with the `self-signed-certificates` charm has been broken. 

Here is an extract of the error: 
```
unit-traefik-public-0: 12:23:02 ERROR unit.traefik-public/0.juju-log certificates:1: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 1032, in <module>
    main(TraefikIngressCharm, use_juju_for_storage=True)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/main.py", line 456, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 1824, in _on_relation_changed
    self.on.certificate_available.emit(
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/observability_libs/v0/cert_handler.py", line 305, in _on_certificate_available
    self.on.cert_changed.emit()  # pyright: ignore
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 532, in wrapped_function
    return callable(*args, **kwargs)  # type: ignore
  File "./src/charm.py", line 378, in _on_cert_changed
    self._update_cert_configs()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 532, in wrapped_function
    return callable(*args, **kwargs)  # type: ignore
  File "./src/charm.py", line 384, in _update_cert_configs
    self.traefik.update_cert_configuration(
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 154, in update_cert_configuration
    self.update_ca_certs()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 195, in update_ca_certs
    self.restart()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 578, in restart
    self._container.restart(self.service_name)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/model.py", line 2136, in restart
    self._pebble.restart_services(service_names)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/pebble.py", line 1962, in restart_services
    return self._services_action('restart', services, timeout, delay)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/pebble.py", line 1983, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "traefik" (cannot start service: exited quickly with code 1)
----- Logs from task 0 -----
2024-04-11T10:23:02Z INFO Service "traefik" has never been started.
----- Logs from task 1 -----
2024-04-11T10:23:02Z INFO Most recent service output:
    2024/04/11 10:23:02 command /bin/traefik error: field not found, node: [0]
2024-04-11T10:23:02Z ERROR cannot start service: exited quickly with code 1
-----
unit-traefik-public-0: 12:23:02 ERROR juju.worker.uniter.operation hook "certificates-relation-changed" (via hook dispatching script: dispatch) failed: exit status 1

```

It seems like there might be a syntax error somewhere in the traefik's yaml configuration. 

### To Reproduce

1. `juju deploy traefik-k8s traefik --channel=latest/edge --revision=174`
2. `juju deploy self-signed-certificates --channel=latest/edge --revision=127`
3. `juju relate traefik:certificates self-signed-certificates`

### Environment

latest/stable, revision 174

### Relevant log output

```shell
unit-traefik-public-0: 12:23:02 ERROR unit.traefik-public/0.juju-log certificates:1: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 1032, in <module>
    main(TraefikIngressCharm, use_juju_for_storage=True)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/main.py", line 456, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tls_certificates_interface/v2/tls_certificates.py", line 1824, in _on_relation_changed
    self.on.certificate_available.emit(
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/observability_libs/v0/cert_handler.py", line 305, in _on_certificate_available
    self.on.cert_changed.emit()  # pyright: ignore
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 865, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 532, in wrapped_function
    return callable(*args, **kwargs)  # type: ignore
  File "./src/charm.py", line 378, in _on_cert_changed
    self._update_cert_configs()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 532, in wrapped_function
    return callable(*args, **kwargs)  # type: ignore
  File "./src/charm.py", line 384, in _update_cert_configs
    self.traefik.update_cert_configuration(
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 154, in update_cert_configuration
    self.update_ca_certs()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 195, in update_ca_certs
    self.restart()
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/src/traefik.py", line 578, in restart
    self._container.restart(self.service_name)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/model.py", line 2136, in restart
    self._pebble.restart_services(service_names)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/pebble.py", line 1962, in restart_services
    return self._services_action('restart', services, timeout, delay)
  File "/var/lib/juju/agents/unit-traefik-public-0/charm/venv/ops/pebble.py", line 1983, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "traefik" (cannot start service: exited quickly with code 1)
----- Logs from task 0 -----
2024-04-11T10:23:02Z INFO Service "traefik" has never been started.
----- Logs from task 1 -----
2024-04-11T10:23:02Z INFO Most recent service output:
    2024/04/11 10:23:02 command /bin/traefik error: field not found, node: [0]
2024-04-11T10:23:02Z ERROR cannot start service: exited quickly with code 1
-----
unit-traefik-public-0: 12:23:02 ERROR juju.worker.uniter.operation hook "certificates-relation-changed" (via hook dispatching script: dispatch) failed: exit status 1

```
```


### Additional context

_No response_

**Comment by PietroPasotti** [comment]
thanks, I'll take a look

**Comment by PietroPasotti** [comment]
Looks like I can't reproduce on my localhost. 
![image](https://github.com/canonical/traefik-k8s-operator/assets/6230162/0e0818f7-a1b8-4441-9c68-eb27cbcd1202)

Any additional information you can add? Platform? Juju version?

**Comment by Thanhphan1147** [comment]
`juju status` showed traefik restarting and I can see the error in debug-log:
```
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "traefik" (cannot start service: exited quickly with code 1)
----- Logs from task 1 -----
2024-04-11T13:16:24Z INFO Most recent service output:
    (...)
    time="2024-04-11T13:16:22Z" level=debug msg="Creating middleware" entryPointName=diagnostics middlewareName=metrics-entrypoint middlewareType=Metrics
    time="2024-04-11T13:16:22Z" level=debug msg="Creating middleware" middlewareType=Metrics entryPointName=web middlewareName=metrics-entrypoint
    time="2024-04-11T13:16:22Z" level=debug msg="Creating middleware" middlewareType=Metrics entryPointName=websecure middlewareName=metrics-entrypoint
    time="2024-04-11T13:16:24Z" level=info msg="I have to go..."
    time="2024-04-11T13:16:24Z" level=info msg="Stopping server gracefully"
    time="2024-04-11T13:16:24Z" level=debug msg="Waiting 10s seconds before killing connections." entryPointName=websecure
    time="2024-04-11T13:16:24Z" level=debug msg="Waiting 10s seconds before killing connections." entryPointName=web
    time="2024-04-11T13:16:24Z" level=error msg="accept tcp [::]:443: use of closed network connection" entryPointName=websecure
    time="2024-04-11T13:16:24Z" level=error msg="close tcp [::]:443: use of closed network connection" entryPointName=websecure
    time="2024-04-11T13:16:24Z" level=debug msg="Entry point websecure closed" entryPointName=websecure
    time="2024-04-11T13:16:24Z" level=debug msg="Waiting 10s seconds before killing connections." entryPointName=diagnostics
    time="2024-04-11T13:16:24Z" level=error msg="accept tcp [::]:80: use of closed network connection" entryPointName=web
    time="2024-04-11T13:16:24Z" level=error msg="close tcp [::]:80: use of closed network connection" entryPointName=web
    time="2024-04-11T13:16:24Z" level=error msg="accept tcp [::]:8082: use of closed network connection" entryPointName=diagnostics
    time="2024-04-11T13:16:24Z" level=error msg="close tcp [::]:8082: use of closed network connection" entryPointName=diagnostics
    time="2024-04-11T13:16:24Z" level=debug msg="Entry point web closed" entryPointName=web
    time="2024-04-11T13:16:24Z" level=debug msg="Entry point diagnostics closed" entryPointName=diagnostics
    time="2024-04-11T13:16:24Z" level=info msg="Server stopped"
    time="2024-04-11T13:16:24Z" level=info msg="Shutting down"
    2024/04/11 13:16:24 command /bin/traefik error: field not found, node: [0]
2024-04-11T13:16:24Z ERROR cannot start service: exited quickly with code 1
-----

```

**Comment by Thanhphan1147** [comment]
I teared down the environment and redeploy the 2 charms and weirdly enough the error didn't show this time ....

**Comment by facundofc** [comment]
In case it helps, I'm pretty sure [this](https://github.com/canonical/traefik-k8s-operator/commit/41683ed3eea8b2e261a7443cf0a96abe1b816d58) introduced the issue.

**Comment by PietroPasotti** [comment]
It might be related, but as far as I can tell the syntax is correct. 
If you manage to reproduce it again, can you share `juju ssh --container traefik traefik/0 cat /etc/traefik/traefik.yaml`?

**Comment by PietroPasotti** [comment]
ah! this might be something. 'redirections' needs to be a dict, not a list (unsure about the yaml terminology)

should look like:
![image](https://github.com/canonical/traefik-k8s-operator/assets/6230162/89eff649-1577-4512-85b5-cf2e49013271)

looks like:
![image](https://github.com/canonical/traefik-k8s-operator/assets/6230162/702000e8-8310-4494-9f8d-08597271a61e)

still, weird that it only errors out sometimes?

**Comment by Thanhphan1147** [comment]
I did teardown/deploy multiple times since and I couldn't reprduce it in the same environment. here is `traefik.yaml` when there's **NO** error :
```
ubuntu@ubuntu:~$ juju ssh --container traefik traefik-k8s/0 cat /etc/traefik/traefik.yaml
entryPoints:
  diagnostics:
    address: :8082
  web:
    address: :80
    http:
    - redirections:
        entryPoint:
          scheme: https
          to: websecure
  websecure:
    address: :443
log:
  level: DEBUG
metrics:
  prometheus:
    addRoutersLabels: true
    addServicesLabels: true
    entryPoint: diagnostics
ping:
  entryPoint: diagnostics
providers:
  file:
    directory: /opt/traefik/juju
    watch: true
ubuntu@ubuntu:~$
```

**Comment by Thanhphan1147** [comment]
here's one of our CI that failed with the error : https://github.com/canonical/jenkins-k8s-operator/actions/runs/8639246022/job/23685277058?pr=140

**Comment by PietroPasotti** [comment]
ah yes, so in the revision that's giving trouble, the code is:
![image](https://github.com/canonical/traefik-k8s-operator/assets/6230162/46794d1d-5e82-4e2e-bc13-50960df3823c)

which generates the wrong config

on `main`, that's been fixed already by @sed-i at 4d3ccc6160d0404601bfe995cdeb7b86d2d7f254

so I guess that's a matter of using `edge`, waiting until the fix reaches `stable`, or if it's a big issue for you, ask our magnanimous @simskij to speed up the release train.
